content/advent-2018/postmortem-debugging-delve.md

```diff
@@ -0,0 +1,469 @@
+
+```
+// starts a consumer for inChannel
+func (m *Metrics) startInChannelConsumer() {
+    for inMetrics := range m.inChannel {
+   	    // ···
+    }
+}
+```
+
+The function reads values out of the channel and does something with them, one by one. In what possible situations, could the sending to this channel being blocked?
+
+When working with channels, there are only four possible "oopsies", according to Dave Cheney's [Channel Axioms](https://dave.cheney.net/2014/03/19/channel-axioms):
+
+- send to a nil channel block forever
+- receive from a nil channel block forever
+- send to a closed channel panics
+- receive from a closed channel returns the zero value immediately.
+
+"Send to a nil channel block forever" – at first sight, this sounds like something possible. But, after double-checking with the code, `inChannel` is initialised in the `Metrics` constructor. So it can't be nil.
+
+Let’s look at the list of functions we’ve got above. Could this (buffered) channel become full because we've stacked somewhere in `(*Metrics).startInChannelConsumer()`?
+
+As you may notice, there is no `startInChannelConsumer` method in the list at all. But what if we stacked somewhere below the method’s callstack?
+
+Delve provides the start position from where we came to the user location, that is `startLoc` field in the JSON. Search for goroutines whose start location was in `startInChannelConsumer` function:
+
+```
+= jq '.result[] | select(.startLoc.function.name | test("startInChannelConsumer$"))' server-test-1_dlv-rpc-list_goroutines.json
+
+{
+  "id": 20,
+  "currentLoc": {
+    "pc": 4387627,
+    "file": "/usr/local/go/src/runtime/proc.go",
+    "line": 303,
+    "function": {
+      "name": "runtime.gopark",
+      "value": 4387392,
+      "type": 0,
+      "goType": 0,
+      "optimized": true
+    }
+  },
+  "userCurrentLoc": {
+    "pc": 7355276,
+    "file": "/vagrant/example/server/metrics.go",
+    "line": 145,
+    "function": {
+      "name": "main.(*Metrics).SetM",
+      "value": 7354992,
+      "type": 0,
+      "goType": 0,
+      "optimized": true
+    }
+  },
+  "goStatementLoc": {
+    "pc": 7354080,
+    "file": "/vagrant/example/server/metrics.go",
+    "line": 95,
+    "function": {
+      "name": "main.NewMetrics",
+      "value": 7353136,
+      "type": 0,
+      "goType": 0,
+      "optimized": true
+    }
+  },
+  "startLoc": {
+    "pc": 7355584,
+    "file": "/vagrant/example/server/metrics.go",
+    "line": 167,
+    "function": {
+      "name": "main.(*Metrics).startInChannelConsumer",
+      "value": 7355584,
+      "type": 0,
+      "goType": 0,
+      "optimized": true
+    }
+  },
+  "threadID": 0
+}
+```
+
+There is a single item in the response. That's promising!
+
+A goroutine with id "20" started at `main.(*Metrics).startInChannelConsumer:167` and went up to `main.(*Metrics).SetM:145` (`userCurrentLoc` field) until it got stacked.
```

Comment:
I'm not familiar with the use of "stacked" in this article. Perhaps "blocked" is better.